File: apps/api/src/geo/geodesy.ts
// geographiclib-geodesic ships CommonJS: under real ESM a named import throws at runtime
// ("does not provide an export named 'Geodesic'"), even though the types permit it and
// bundler-based test runners paper over it. Import the default and destructure.
import geographiclib from 'geographiclib-geodesic';
import type { Feature, LineString, Point, Polygon, Position } from 'geojson';

const { Geodesic } = geographiclib;
import { area as turfArea } from '@turf/turf';
import { km, metres, sqKm, type Kilometres, type Metres, type SquareKm } from '@varuna/shared';

/**
 * Ellipsoidal geodesy on WGS84 via GeographicLib (Karney). `pyproj.Geod` in the Python
 * service uses the same algorithm — the known-answer suite (geodesy.test.ts) proves the two
 * stacks agree within 0.1% (02_TRD §2.6.4 / §2.15).
 *
 * Rules:
 *  - Distance / length / area here are ALWAYS ellipsoidal, never in degrees (02_TRD TR-3).
 *  - Every function takes and returns branded units. Bare `number` for a distance or area
 *    is rejected in review (05_FRONTEND §5.2, 02_TRD §2.6.4).
 *  - `$near` is never used against a polygon — that lint rule lives in eslint.config.mjs.
 */
const geod = Geodesic.WGS84;

/** Shortest ellipsoidal distance between two [lon, lat] points, in metres. */
export function geodesicDistanceM(
  a: readonly [number, number],
  b: readonly [number, number],
): Metres {
  const r = geod.Inverse(a[1], a[0], b[1], b[0]);
  return metres(r.s12 ?? 0);
}

export function geodesicDistanceKm(
  a: readonly [number, number],
  b: readonly [number, number],
): Kilometres {
  return km((geodesicDistanceM(a, b) as number) / 1000);
}

/** Initial bearing (degrees true, [0,360)) from `a` to `b`. */
export function geodesicBearingDeg(
  a: readonly [number, number],
  b: readonly [number, number],
): number {
  const r = geod.Inverse(a[1], a[0], b[1], b[0]);
  return (((r.azi1 ?? 0) % 360) + 360) % 360;
}

/** Ellipsoidal area of a polygon (outer ring minus holes), in m². */
export function geodesicPolygonAreaM2Value(polygon: Polygon): number {
  let total = 0;
  polygon.coordinates.forEach((ring, i) => {
    const poly = geod.Polygon(false);
    for (const pos of ringWithoutClosingDuplicate(ring)) {
      poly.AddPoint(Number(pos[1]), Number(pos[0]));
    }
    const res = poly.Compute(false, true);
    const ringArea = Math.abs(res.area ?? 0);
    total += i === 0 ? ringArea : -ringArea;
  });
  return total;
}

export function geodesicPolygonAreaKm2(polygon: Polygon): SquareKm {
  return sqKm(geodesicPolygonAreaM2Value(polygon) / 1e6);
}

/** Geodesic length of a LineString, in km. */
export function geodesicLengthKm(line: LineString): Kilometres {
  let total = 0;
  const c = line.coordinates;
  for (let i = 1; i < c.length; i++) {
    const prev = c[i - 1];
    const cur = c[i];
    if (!prev || !cur) continue;
    total += geodesicDistanceM([prev[0]!, prev[1]!], [cur[0]!, cur[1]!]) as number;
  }
  return km(total / 1000);
}

function ringWithoutClosingDuplicate(ring: Position[]): Position[] {
  if (ring.length < 2) return ring;
  const first = ring[0]!;
  const last = ring[ring.length - 1]!;
  return first[0] === last[0] && first[1] === last[1] ? ring.slice(0, -1) : ring;
}

/**
 * Planar/spherical area via Turf — used ONLY where a fast approximate area is acceptable
 * (e.g. the AOI size guard). For anything that becomes evidence, use
 * `geodesicPolygonAreaKm2`.
 */
export function approxPolygonAreaKm2(polygon: Polygon | Feature<Polygon>): SquareKm {
  return sqKm(turfArea(polygon as Feature<Polygon>) / 1e6);
}

export type { Point, Polygon, LineString, Feature };

File: services/ml/varuna_ml/geo/geodesy.py
"""Ellipsoidal geodesy on WGS84 via pyproj.Geod (PROJ's GeographicLib implementation).

Must agree with the Node stack (apps/api/src/geo/geodesy.ts, geographiclib-geodesic) within
0.1% — the known-answer suite in tests/test_geodesy_known_answers.py enforces this against
the shared packages/shared/geo-known-answers.json (02_TRD §2.6.4 / §2.15).

Coordinates are [longitude, latitude] everywhere, matching the rest of the system
(02_TRD TR-2). Area and distance are NEVER computed in degrees (02_TRD TR-3).
"""

from __future__ import annotations

from pyproj import Geod
from shapely.geometry import Polygon, mapping
from shapely.geometry.base import BaseGeometry

_GEOD = Geod(ellps="WGS84")


def geodesic_distance_m(a: tuple[float, float], b: tuple[float, float]) -> float:
    """Shortest ellipsoidal distance between two (lon, lat) points, in metres."""
    _, _, dist = _GEOD.inv(a[0], a[1], b[0], b[1])
    return abs(dist)


def geodesic_bearing_deg(a: tuple[float, float], b: tuple[float, float]) -> float:
    """Initial bearing from a to b, degrees true in [0, 360)."""
    fwd_az, _, _ = _GEOD.inv(a[0], a[1], b[0], b[1])
    return fwd_az % 360.0


def geodesic_polygon_area_m2(ring_lonlat: list[tuple[float, float]]) -> float:
    """Ellipsoidal area of a single ring given as (lon, lat) vertices, in m²."""
    lons = [p[0] for p in ring_lonlat]
    lats = [p[1] for p in ring_lonlat]
    area, _perimeter = _GEOD.polygon_area_perimeter(lons, lats)
    return abs(area)


def geodesic_area_perimeter(geom: BaseGeometry) -> tuple[float, float]:
    """Ellipsoidal (area_m2, perimeter_m) of a shapely Polygon in EPSG:4326.

    Holes are subtracted. Used for slick polygons — the value that becomes evidence
    (07_AIML §7.2.10).
    """
    if not isinstance(geom, Polygon):
        raise TypeError(f"expected Polygon, got {geom.geom_type}")
    gj = mapping(geom)
    rings = gj["coordinates"]
    total_area = 0.0
    total_per = 0.0
    for i, ring in enumerate(rings):
        lons = [c[0] for c in ring]
        lats = [c[1] for c in ring]
        a, p = _GEOD.polygon_area_perimeter(lons, lats)
        total_area += abs(a) if i == 0 else -abs(a)
        total_per += abs(p)
    return total_area, total_per


def geodesic_line_length_km(coords_lonlat: list[tuple[float, float]]) -> float:
    """Geodesic length of a polyline given as (lon, lat) vertices, in km."""
    total = 0.0
    for prev, cur in zip(coords_lonlat, coords_lonlat[1:]):
        total += geodesic_distance_m(prev, cur)
    return total / 1000.0
